R/source.R

```diff
@@ -429,33 +434,22 @@ check_cargo_output <- function(compilation_result, message_buffer, tty_has_color
     jsonlite::parse_json
   )
 
-  if (!isTRUE(quiet)) {
-    purrr::walk(
-      gather_cargo_output(
-        cargo_output,
-        "warning",
-        tty_has_colors
-      ),
-      ui_w
-    )
-  }
-
   if (!isTRUE(compilation_result$status == 0)) {
-    error_messages <- purrr::map_chr(
-      gather_cargo_output(
-        cargo_output,
-        "error",
-        tty_has_colors
-      ),
-      bullet_x
+
+    error_messages <- gather_cargo_output(
+      cargo_output,
+      "error",
+      tty_has_colors
     )
 
-    ui_throw(
-      "Rust code could not be compiled successfully. Aborting.",
-      error_messages,
-      call = call,
-      glue_open = "{<{",
-      glue_close = "}>}"
+    cli::cli_abort(
+      cli::cli_fmt({
+      cli::cli_text("Rust code could not be compiled successfully. Aborting.")
+
+      if (!quiet) purrr::walk(error_messages, cli::cli_alert_danger)
```

Comment:
This is odd. Do you not want to `purrr::map` here? Or `walk` outside of the abort?

Reply:
Per @Ilia-Kosenkov's comment above. This is a surprisingly tough problem! `walk` is preferable to `map` here since we are using `cli::cli_alert_danger` for its side effects (printing to the console) rather than collecting the output from it. 

`error_messages` also needs to be iterated over otherwise the errors are collapsed and `\n`s are removed. 

Though with the addition of `withr::local_options("cli.default_handler")` I think the `if (!quiet)` can be removed.